Support new function pointer metadata encoding

333fred opened the pull request:
Adds support for the new extensible calling convention proposal in metadata. We bump the version of SRM (and therefore SCI) in order to be able to read the new calling convention bit, and then implement support for looking through the correct set of custom modifiers to be able to check calling convention equivalence. Order and duplication among the conventions does not matter for symbol equality. Modiifer order is copied when overriding as usual. For now, we have a flag to override checking for the RuntimeFeatures flag, as we don't yet have a runtime to test it on.

333fred commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -6259,23 +6257,118 @@ unsafe class Base
+                        Assert.True(baseM1.Parameters.Single().Type.Equals(derivedM1.Parameters.Single().Type, TypeCompareKind.ConsiderEverything));
+                        Assert.True(baseM2.ReturnType.Equals(derivedM2.ReturnType, TypeCompareKind.ConsiderEverything));
+                        Assert.True(baseM3.Parameters.Single().Type.Equals(derivedM3.Parameters.Single().Type, TypeCompareKind.ConsiderEverything));
+                        Assert.True(baseM3.ReturnType.Equals(derivedM3.ReturnType, TypeCompareKind.ConsiderEverything));
```
Needs to be M4

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -3263,7 +3263,16 @@ protected override IPointerTypeSymbol CommonCreatePointerTypeSymbol(ITypeSymbol
                 type => TypeWithAnnotations.Create(type.EnsureCSharpSymbolOrNull(nameof(parameterTypes)), type.NullableAnnotation.ToInternalAnnotation()));
 
-            return FunctionPointerTypeSymbol.CreateFromParts(returnTypeWithAnnotations, returnRefKind, parameterTypesWithAnnotations, parameterRefKinds, this).GetPublicSymbol();
+            // PROTOTYPE(func-ptr): Allow passing in calling convention and ref custom modifiers for return/parameters
```
>ref custom modifiers [](start = 76, length = 20)

"ref custom modifiers"?

333fred replied:
What's the question?

---
In reply to: [457680708](https://github.com/dotnet/roslyn/pull/46110#discussion_r457680708) [](ancestors = 457680708)

AlekseyTs replied:
> What's the question?

Why only ref modifiers, it looks like we are not allowing to pass any modifiers

---
In reply to: [457779909](https://github.com/dotnet/roslyn/pull/46110#discussion_r457779909) [](ancestors = 457779909,457680708)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -1561,7 +1561,7 @@ private void CheckMemberNameConflicts(DiagnosticBag diagnostics)
                     //
 
-                    if ((object)lastSym != null)
+                    if ((object?)lastSym != null)
```
>(object?)lastSym != null [](start = 24, length = 24)

```lastSym is object```? Since you are changing this line.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -7,8 +7,10 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Globalization;
```
>Globalization [](start = 13, length = 13)

Why is this needed?

333fred replied:
For `CultureInfo`

---
In reply to: [457684561](https://github.com/dotnet/roslyn/pull/46110#discussion_r457684561) [](ancestors = 457684561)

AlekseyTs replied:
> For CultureInfo

I think we should be using binary comparison option instead and simply concatenating strings instead of using ```string.Format``, if concatenation cannot be avoided. 

---
In reply to: [457780357](https://github.com/dotnet/roslyn/pull/46110#discussion_r457780357) [](ancestors = 457780357,457684561)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -19,14 +21,16 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
             RefKind refKind = RefKind.None;
             TypeWithAnnotations returnType;
-            var refReadonlyModifiers = ImmutableArray<CustomModifier>.Empty;
+            var refCustomModifiers = ImmutableArray<CustomModifier>.Empty;
```
>var refCustomModifiers = ImmutableArray<CustomModifier>.Empty; [](start = 12, length = 62)

Consider moving this declaration closer to where this local can get a different value assigned.

333fred replied:
I can't, it needs to be on this level for use in the final `if`. This is as close as it can get.

---
In reply to: [457686746](https://github.com/dotnet/roslyn/pull/46110#discussion_r457686746) [](ancestors = 457686746)

AlekseyTs replied:
> I can't, it needs to be on this level for use in the final if. This is as close as it can get.

It looks like the first reference to this local, ```refCustomModifiers```, is on line 93. It feels like the declaration can be moved to line 90.


---
In reply to: [457780486](https://github.com/dotnet/roslyn/pull/46110#discussion_r457780486) [](ancestors = 457780486,457686746)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +123,100 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
+                    if (!typeBinder.Compilation.Assembly.RuntimeSupportsUnmanagedSignatureCallingConvention)
                     {
-                        diagnostics.Add(ErrorCode.ERR_InvalidFunctionPointerCallingConvention, node.GetLocation(), givenConvention);
+                        diagnostics.Add(isExtensionError ? ErrorCode.ERR_RuntimeDoesNotSupportUnmanagedExtensionCallConv
```
>ERR_RuntimeDoesNotSupportUnmanagedExtensionCallConv [](start = 69, length = 51)

I am not sure I see a benefit from differentiating the error. Since we check the same runtime support flag, the both scenarios are either supported, or not supported at the same time.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +123,100 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
+                    string specifierText = specifier.Name.ValueText;
+                    var metadataName = string.Format(CultureInfo.InvariantCulture, CallingConventionMetadataFormatString, specifierText);
+
+                    var specifierType = typeBinder.Compilation.Assembly.CorLibrary.GetTypeByMetadataName(metadataName, includeReferences: false, isWellKnownType: true, conflicts: out _, warnings: null);
```
>GetTypeByMetadataName [](start = 83, length = 21)

I do not think we ever use this API internally, it is purely to support tests and public API consumers. I think we should call ```internal NamedTypeSymbol LookupTopLevelMetadataType(ref MetadataTypeName emittedName, bool digThroughForwardedTypes)``` API directly. That should give us several benefits:
1) No need to build full type name with  ```string.Format``` and then immediately decompose it.
2) I do not believe it ever return null, but rather returns a "good" error type symbol that preserves the target type's identity.
3) I think that we can simply report its use-site diagnostics to give the user specific error message about what type we failed to locate, rather than reporting a misleading error: "'{0}' is not a valid calling convention specifier for a function pointer. Valid conventions are 'Cdecl', 'Stdcall', 'Thiscall', and 'Fastcall'."

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +123,100 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
+                        diagnostics.Add(ErrorCode.ERR_InvalidFunctionPointerCallingConvention, specifier.GetLocation(), specifierText);
+                    }
+
+                    return CSharpCustomModifier.CreateOptional(specifierType);
```
>specifierType [](start = 63, length = 13)

A use-site diagnostics must be reported for the type.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +123,100 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
+                    string specifierText = specifier.Name.ValueText;
+                    var metadataName = string.Format(CultureInfo.InvariantCulture, CallingConventionMetadataFormatString, specifierText);
+
+                    var specifierType = typeBinder.Compilation.Assembly.CorLibrary.GetTypeByMetadataName(metadataName, includeReferences: false, isWellKnownType: true, conflicts: out _, warnings: null);
```
>true [](start = 162, length = 4)

This is not a well known type

333fred replied:
Yes, but I believe we want it to behave like a well-known type for all intents and purposes. This flag checks that the symbol is accessible and public.

---
In reply to: [457723185](https://github.com/dotnet/roslyn/pull/46110#discussion_r457723185) [](ancestors = 457723185)

AlekseyTs replied:
> Yes, but I believe we want it to behave like a well-known type for all intents and purposes. This flag checks that the symbol is accessible and public.

We do want them to be public and we can check for that. We do not want them to be treated as well-known even if at the moment the requirements appear to match. Also, in my other comment I mentioned that I think we should use a different lookup API, that provides guarantees that we would want. 

---
In reply to: [457785546](https://github.com/dotnet/roslyn/pull/46110#discussion_r457785546) [](ancestors = 457785546,457723185)

333fred replied:
> Also, in my other comment I mentioned that I think we should use a different lookup API, that provides guarantees that we would want. 

I don't believe it does: this codepath uses the API you called out, and then explicitly checks the return type to ensure it's public and visible.

---
In reply to: [458310033](https://github.com/dotnet/roslyn/pull/46110#discussion_r458310033) [](ancestors = 458310033,457785546,457723185)

AlekseyTs replied:
> I don't believe it does: this codepath uses the API you called out, and then explicitly checks the return type to ensure it's public and visible.

The guarantees that I refer to are mentioned in that other comment, they are not mentioned on this thread. If you are referring to the public accessibility check, I think we want to perform this check here explicitly and report good error that the type such-and-such is not public instead of trying to figure out why ```GetTypeByMetadataName``` failed to locate the type .

---
In reply to: [458369877](https://github.com/dotnet/roslyn/pull/46110#discussion_r458369877) [](ancestors = 458369877,458310033,457785546,457723185)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +123,100 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
+                    if (specifierType is null)
+                    {
+                        // Type name is CallConv{0}
+                        specifierType = typeBinder.CreateErrorType(name: metadataName[32..]);
```
>32 [](start = 86, length = 2)

Magic constant?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -84,18 +88,26 @@ public static FunctionPointerMethodSymbol CreateFromSource(FunctionPointerTypeSy
 
-            // PROTOTYPE(func-ptr): Do this right. Just making compatible changes for now
-            static CallingConvention getCallingConvention(FunctionPointerCallingConventionSyntax? callingConventionSyntax, DiagnosticBag diagnostics)
+            static CallingConvention getCallingConvention(Binder typeBinder, FunctionPointerCallingConventionSyntax? callingConventionSyntax, ArrayBuilder<CustomModifier> customModifiers, DiagnosticBag diagnostics)
```
>typeBinder [](start = 65, length = 10)

It looks like what we need is the compilation. Having dependency on the binder feels strange, there is nothing to "bind" in any particular scope or context here.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -333,17 +402,24 @@ public FunctionPointerMethodSymbol SetNullabilityForReferenceTypes(Func<TypeWith
             ReturnTypeWithAnnotations = returnTypeWithAnnotations;
-            _parameters = parameterTypes.ZipAsArray(parameterRefKinds, (Method: this, Comp: compilation), (type, refKind, i, arg) =>
-                new FunctionPointerParameterSymbol(type, refKind, i, arg.Method, refCustomModifiers: getCustomModifierForRefKind(refKind, arg.Comp)));
+            _parameters = parameterTypes.ZipAsArray(parameterRefKinds, (Method: this, Comp: compilation, ParamCustomModifiers: parameterRefCustomModifiers),
```
>ParamCustomModifiers [](start = 105, length = 20)

"ParamRefCustomModifiers"?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -333,17 +402,24 @@ public FunctionPointerMethodSymbol SetNullabilityForReferenceTypes(Func<TypeWith
             Debug.Assert(refKind != RefKind.Out);
-            RefCustomModifiers = getCustomModifierForRefKind(refKind, compilation);
+            Debug.Assert(parameterRefCustomModifiers.IsDefault || parameterRefCustomModifiers.Length == parameterTypes.Length);
+            RefCustomModifiers = refCustomModifiers.IsDefault ? getCustomModifierForRefKind(refKind, compilation) : refCustomModifiers;
```
>refCustomModifiers [](start = 116, length = 18)

I assume this is the constructor to support public API for creating the symbol. Do you expect the user be responsible for adding In/Out modifiers when they want to specify calling convention modifiers?

333fred replied:
I'll leave a prototype comment. I'm currently unsure, I want to talk with Jason/Cyrus about that.

---
In reply to: [457729603](https://github.com/dotnet/roslyn/pull/46110#discussion_r457729603) [](ancestors = 457729603)

AlekseyTs commented:
>         private FunctionPointerMethodSymbol(

Are changes in this function covered by tests?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:401 in e2f6eff. [](commit_id = e2f6effc65a4cecb070dee40365236ccec4fe97a, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -483,6 +559,62 @@ internal FunctionPointerMethodSymbol ApplyNullableTransforms(byte defaultTransfo
+
+                    if (builder.Count == 0)
+                    {
+                        Interlocked.CompareExchange(ref _lazyCallingConventionModifiers, ImmutableHashSet<CustomModifier>.Empty, null);
```
>ImmutableHashSet<CustomModifier> [](start = 89, length = 32)

Is this a reference type? We don't really need to do ```Interlocked.CompareExchange``` then.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -483,6 +559,62 @@ internal FunctionPointerMethodSymbol ApplyNullableTransforms(byte defaultTransfo
+            return _lazyCallingConventionModifiers;
+        }
+
+        private static bool IsCallingConventionModifier(CustomModifier modifier)
```
>IsCallingConventionModifier [](start = 28, length = 27)

What is this function going to return for type named "System.Runtime.CompilerServices.CallConv"? Do we have a test for a scenario with a modifier like that?

333fred replied:
Added `CallingConventionNamedCallConv`

---
In reply to: [457733128](https://github.com/dotnet/roslyn/pull/46110#discussion_r457733128) [](ancestors = 457733128)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +123,100 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
+                static CustomModifier handleIndividualUnrecognizedSpecifier(FunctionPointerUnmanagedCallingConventionSyntax specifier, Binder typeBinder, DiagnosticBag diagnostics)
+                {
+                    const string CallingConventionMetadataFormatString = "System.Runtime.CompilerServices.CallConv{0}";
+                    string specifierText = specifier.Name.ValueText;
```
>specifierText [](start = 27, length = 13)

We probably want to return if this is an empty string.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +123,100 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
-                                return reportBadConventionAndReturn(givenConvention: "", unmanagedList, diagnostics);
+                                if (!unmanagedList.ContainsDiagnostics)
+                                {
+                                    diagnostics.Add(ErrorCode.ERR_InvalidFunctionPointerCallingConvention, unmanagedList.OpenBracketToken.GetLocation(), "");
```
>ERR_InvalidFunctionPointerCallingConvention [](start = 62, length = 43)

Is this message going to be confusing?

333fred replied:
Maybe, but as the comment says you can't observe this from parser-created code so I don't think diagnostic quality is particularly important.

---
In reply to: [457734288](https://github.com/dotnet/roslyn/pull/46110#discussion_r457734288) [](ancestors = 457734288)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -268,36 +282,61 @@ class C
+                // (6,47): error CS9502: The target runtime doesn't support extensible unmanaged calling conventions.
+                //     public unsafe void M3(delegate* unmanaged[]<void> p) {}
+                Diagnostic(ErrorCode.ERR_RuntimeDoesNotSupportUnmanagedExtensionCallConv, "").WithLocation(6, 47),
+                // (6,47): error CS8807: '' is not a valid calling convention specifier for a function pointer. Valid conventions are 'Cdecl', 'Stdcall', 'Thiscall', and 'Fastcall'.
```
>Valid conventions are 'Cdecl', 'Stdcall', 'Thiscall', and 'Fastcall'. [](start = 112, length = 69)

This feels misleading

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -268,36 +282,61 @@ class C
+
+            FunctionPointerUtilities.VerifyFunctionPointerSemanticInfo(model, functionPointers[2],
+                expectedSyntax: "delegate* unmanaged[]<void>",
+                expectedType: "delegate*<System.Void modopt(CallConv)>",
```
>modopt(CallConv) [](start = 53, length = 16)

This doesn't look right. We probably shouldn't add any modifiers for this syntax error case.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -6259,23 +6257,118 @@ unsafe class Base
+            var allSourceComp = CreateCompilationWithFunctionPointers(source1 + source2);
+
+            allSourceComp.Assembly.SetOverrideRuntimeSupportsUnmanagedSignatureCallingConvention();
+            CompileAndVerify(allSourceComp, symbolValidator: getSymbolValidator(separateAssembly: false));
```
>CompileAndVerify [](start = 12, length = 16)

When it becomes possible, we should observe runtime behavior, i.e. things are overridden as we expect them to be.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -6259,23 +6257,118 @@ unsafe class Base
+
+            var derivedComp = CreateCompilationWithFunctionPointers(source2, references: new[] { metadataRef });
+            derivedComp.Assembly.SetOverrideRuntimeSupportsUnmanagedSignatureCallingConvention();
+            CompileAndVerify(derivedComp, symbolValidator: getSymbolValidator(separateAssembly: true));
```
>CompileAndVerify [](start = 12, length = 16)

Same comment

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -6259,23 +6257,118 @@ unsafe class Base
+                        var derivedM4 = derived.GetMethod("M4");
+
+                        Assert.True(baseM1.Parameters.Single().Type.Equals(derivedM1.Parameters.Single().Type, TypeCompareKind.ConsiderEverything));
+                        Assert.True(baseM2.ReturnType.Equals(derivedM2.ReturnType, TypeCompareKind.ConsiderEverything));
```
>ReturnType [](start = 43, length = 10)

It looks like this doesn't involve modifiers

333fred replied:
It involved the modifiers in the actual function pointer type itself. I can make it assert the modifiers after the full function pointer type itself, but that's not the interesting part of this test.

---
In reply to: [457764492](https://github.com/dotnet/roslyn/pull/46110#discussion_r457764492) [](ancestors = 457764492)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -6259,23 +6257,118 @@ unsafe class Base
+                        var derivedM4 = derived.GetMethod("M4");
+
+                        Assert.True(baseM1.Parameters.Single().Type.Equals(derivedM1.Parameters.Single().Type, TypeCompareKind.ConsiderEverything));
+                        Assert.True(baseM2.ReturnType.Equals(derivedM2.ReturnType, TypeCompareKind.ConsiderEverything));
```
>ReturnType [](start = 43, length = 10)

We probably want to assert Ref modifiers as well.

333fred replied:
Same comment. The ref modifiers of the function pointer type itself are asserted.

---
In reply to: [457764703](https://github.com/dotnet/roslyn/pull/46110#discussion_r457764703) [](ancestors = 457764703)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -7454,6 +7547,30 @@ .maxstack  1
+    unsafe void M()
+    {
+        delegate* unmanaged<void> ptr1;
+        delegate* unmanaged[Stdcall, Thiscall]<void> ptr2;
```
>delegate* unmanaged[Stdcall, Thiscall]<void> ptr2; [](start = 8, length = 50)

I think we want to verify the shape of the symbol for scenario when runtime doesn't support the feature.

AlekseyTs commented:
It feels like not all the affected code paths are covered by tests. Should test error scenarios when types cannot be found etc.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -268,36 +282,61 @@ class C
+                // (4,47): error CS9502: The target runtime doesn't support extensible unmanaged calling conventions.
+                //     public unsafe void M1(delegate* unmanaged[invalid]<void> p) {}
+                Diagnostic(ErrorCode.ERR_RuntimeDoesNotSupportUnmanagedExtensionCallConv, "invalid").WithLocation(4, 47),
                 // (4,47): error CS8807: 'invalid' is not a valid calling convention specifier for a function pointer. Valid conventions are 'Cdecl', 'Stdcall', 'Thiscall', and 'Fastcall'.
```
>Valid conventions are 'Cdecl', 'Stdcall', 'Thiscall', and 'Fastcall' [](start = 119, length = 68)

This feels misleading.

333fred replied:
How so?

---
In reply to: [457765833](https://github.com/dotnet/roslyn/pull/46110#discussion_r457765833) [](ancestors = 457765833)

AlekseyTs replied:
> How so?

In a sense is that any identifier is a fair game if we can locate the corresponding type. So, the closed set of valid conventions is misleading in my opinion.

---
In reply to: [457768821](https://github.com/dotnet/roslyn/pull/46110#discussion_r457768821) [](ancestors = 457768821,457765833)

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented:
>         private FunctionPointerMethodSymbol(

Yes. The new equality tests pass non-default arrays of custom modifiers, and existing tests call through CreateFunctionPointer.

---
In reply to: [661392036](https://github.com/dotnet/roslyn/pull/46110#issuecomment-661392036) [](ancestors = 661392036)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:401 in e2f6eff. [](commit_id = e2f6effc65a4cecb070dee40365236ccec4fe97a, deletion_comment = False)

333fred commented:
> Should test error scenarios when types cannot be found etc.

When what types can't be found? I have tests for valid scenarios for when `CallConvCdecl` et all aren't there, and I have tests for when there's no definition in the core library for the given modifer?

---
In reply to: [661486448](https://github.com/dotnet/roslyn/pull/46110#issuecomment-661486448) [](ancestors = 661486448)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -483,6 +559,62 @@ internal FunctionPointerMethodSymbol ApplyNullableTransforms(byte defaultTransfo
             }
         }
 
+        public ImmutableHashSet<CustomModifier> GetCallingConventionModifiers()
```
>GetCallingConventionModifiers [](start = 48, length = 29)

It feels like we want to have tests with:
- duplicate modifiers
- with unrelated modifiers before, after, and in between
- interesting modifiers before and after ref byte at the same time, those after the ref bite shouldn't affect the calling convention matching.

333fred replied:
We have the first two (see `Equality_UnmanagedExtensionModifiers`). I'll add the last one.

---
In reply to: [458297902](https://github.com/dotnet/roslyn/pull/46110#discussion_r458297902) [](ancestors = 458297902)

333fred replied:
Added `Equality_DifferingRefAndTypeCustomModifiers`

---
In reply to: [458447003](https://github.com/dotnet/roslyn/pull/46110#discussion_r458447003) [](ancestors = 458447003,458297902)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -483,6 +559,62 @@ internal FunctionPointerMethodSymbol ApplyNullableTransforms(byte defaultTransfo
+        {
+            var modifierType = ((CSharpCustomModifier)modifier).ModifierSymbol;
+            return (object)modifierType.ContainingAssembly == modifierType.ContainingAssembly.CorLibrary
+                   && modifierType.Name.StartsWith("CallConv", ignoreCase: false, culture: CultureInfo.InvariantCulture)
```
>CultureInfo.InvariantCulture [](start = 91, length = 28)

I think we want binary comparison option instead.

AlekseyTs commented:
> When what types can't be found? I have tests for valid scenarios for when CallConvCdecl et all aren't there, and I have tests for when there's no definition in the core library for the given modifier?

I guess I didn't see any tests with errors that I would expect. Specifically errors saying that the type with calculated name is not found.


---
In reply to: [661527772](https://github.com/dotnet/roslyn/pull/46110#issuecomment-661527772) [](ancestors = 661527772,661486448)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -483,6 +559,62 @@ internal FunctionPointerMethodSymbol ApplyNullableTransforms(byte defaultTransfo
+        private static bool IsCallingConventionModifier(CustomModifier modifier)
+        {
+            var modifierType = ((CSharpCustomModifier)modifier).ModifierSymbol;
+            return (object)modifierType.ContainingAssembly == modifierType.ContainingAssembly.CorLibrary
```
>(object)modifierType.ContainingAssembly == modifierType.ContainingAssembly.CorLibrary [](start = 19, length = 85)

Do we have a test where a modifier has the right name, but wrong containing assembly?

333fred replied:
Added `ConventionDefinedInWrongAssembly`

---
In reply to: [458923114](https://github.com/dotnet/roslyn/pull/46110#discussion_r458923114) [](ancestors = 458923114)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -483,6 +559,62 @@ internal FunctionPointerMethodSymbol ApplyNullableTransforms(byte defaultTransfo
+        {
+            var modifierType = ((CSharpCustomModifier)modifier).ModifierSymbol;
+            return (object)modifierType.ContainingAssembly == modifierType.ContainingAssembly.CorLibrary
+                   && modifierType.Name.StartsWith("CallConv", ignoreCase: false, culture: CultureInfo.InvariantCulture)
```
> modifierType.Name.StartsWith("CallConv", ignoreCase: false, culture: CultureInfo.InvariantCulture) [](start = 21, length = 99)

Do we have a test where a modifier has the right name, but is generic?

333fred replied:
Added `GenericCallingConventionType`.

---
In reply to: [458923748](https://github.com/dotnet/roslyn/pull/46110#discussion_r458923748) [](ancestors = 458923748)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +123,100 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
+                    string specifierText = specifier.Name.ValueText;
+                    var metadataName = string.Format(CultureInfo.InvariantCulture, CallingConventionMetadataFormatString, specifierText);
+
+                    var specifierType = typeBinder.Compilation.Assembly.CorLibrary.GetTypeByMetadataName(metadataName, includeReferences: false, isWellKnownType: true, conflicts: out _, warnings: null);
```
>GetTypeByMetadataName [](start = 83, length = 21)

Do we have a test where there is a type with a matching name, but it is generic?

333fred replied:
Added `GenericCallingConventionType`

---
In reply to: [458925153](https://github.com/dotnet/roslyn/pull/46110#discussion_r458925153) [](ancestors = 458925153)

333fred commented:
I believe the errors should now be what you expect. Let me know if they're not.

---
In reply to: [662039272](https://github.com/dotnet/roslyn/pull/46110#issuecomment-662039272) [](ancestors = 662039272,661527772,661486448)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [661487980](https://github.com/dotnet/roslyn/pull/46110#issuecomment-661487980) [](ancestors = 661487980)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -111,48 +122,119 @@ static CallingConvention getCallingConvention(FunctionPointerCallingConventionSy
+                    }
+
+                    string typeName = "CallConv" + specifierText;
+                    var metadataName = MetadataTypeName.FromNamespaceAndTypeName("System.Runtime.CompilerServices", typeName);
```
>FromNamespaceAndTypeName [](start = 56, length = 24)

I think we want to use ```useCLSCompliantNameArityEncoding: true``` for this name. We don't want to pick up a generic type that is missing the arity suffix in the name. It would be good to test for this.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -7454,6 +7550,296 @@ .maxstack  1
+                parameterRefKinds: ImmutableArray<RefKind>.Empty, comp2);
+            var funcPtrRef = FunctionPointerTypeSymbol.CreateFromParts(
+                CallingConvention.Unmanaged, TypeWithAnnotations.Create(@string), refCustomModifiers: default,
+                returnRefKind: RefKind.None, parameterTypes: ImmutableArray<TypeWithAnnotations>.Empty, parameterRefCustomModifiers: default,
```
>None [](start = 39, length = 4)

Is this supposed to be ```Ref```?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -7454,6 +7550,296 @@ .maxstack  1
+                parameterRefKinds: ImmutableArray<RefKind>.Empty, comp2);
+            var funcPtrWithTestOnRef = FunctionPointerTypeSymbol.CreateFromParts(
+                CallingConvention.Unmanaged, TypeWithAnnotations.Create(@string), refCustomModifiers: ImmutableArray.Create(testMod),
+                returnRefKind: RefKind.None, parameterTypes: ImmutableArray<TypeWithAnnotations>.Empty, parameterRefCustomModifiers: default,
```
>None [](start = 39, length = 4)

Is this supposed to be ```Ref```?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -7454,6 +7550,296 @@ .maxstack  1
+                parameterRefKinds: ImmutableArray<RefKind>.Empty, comp1);
+            var funcPtrRef = FunctionPointerTypeSymbol.CreateFromParts(
+                CallingConvention.Unmanaged, TypeWithAnnotations.Create(@string), refCustomModifiers: default,
+                returnRefKind: RefKind.None, parameterTypes: ImmutableArray<TypeWithAnnotations>.Empty, parameterRefCustomModifiers: default,
```
>None [](start = 39, length = 4)

Is this supposed to be ```Ref```? Check all other similar places. It looks like this is copied around

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -7454,6 +7550,296 @@ .maxstack  1
+                returnRefKind: RefKind.None, parameterTypes: ImmutableArray<TypeWithAnnotations>.Empty, parameterRefCustomModifiers: default,
+                parameterRefKinds: ImmutableArray<RefKind>.Empty, comp1);
+
+            Assert.True(funcPtr.Equals(funcPtrWithTestOnReturn, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds));
```
>Assert.True [](start = 12, length = 11)

Consider calling GetCallingConventionModifiers for funcPtrWithTestOnReturn and funcPtrWithTestOnRef and asserting the set is empty

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -7454,6 +7550,296 @@ .maxstack  1
+                returnRefKind: RefKind.None, parameterTypes: ImmutableArray<TypeWithAnnotations>.Empty, parameterRefCustomModifiers: default,
+                parameterRefKinds: ImmutableArray<RefKind>.Empty, comp2);
+
+            Assert.True(funcPtr.Equals(funcPtrWithTestOnReturn, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds));
```
>Assert.True [](start = 12, length = 11)

Consider calling GetCallingConventionModifiers for funcPtrWithTestOnReturn and funcPtrWithTestOnRef and asserting the set is empty

AlekseyTs replied:
This suggestion applies to other similar tests that use equality as an indirect way to check what modifiers we consider significant.

---
In reply to: [459109795](https://github.com/dotnet/roslyn/pull/46110#discussion_r459109795) [](ancestors = 459109795)

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -1827,5 +1864,208 @@ void M()
+                => (FunctionPointerTypeSymbol.CreateFromParts(
+                        callingConvention,
+                        TypeWithAnnotations.Create(returnType, customModifiers: typeCustomModifiers),
+                        refCustomModifiers: refCustomModifiers,
```
>refCustomModifiers [](start = 44, length = 18)

It feels like this should always be empty. I suggest adding an assert to the CreateFromParts API, or constructor that it uses.

AlekseyTs commented:
Done with review pass (iteration 4)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [662726512](https://github.com/dotnet/roslyn/pull/46110#issuecomment-662726512) [](ancestors = 662726512)

AlekseyTs commented:
It looks like CI is failing

AlekseyTs commented:
Done with review pass (iteration 5)

333fred commented:
@AlekseyTs tests are now passing. @jaredpar for another review pass, since there have been changes.

AlekseyTs commented on src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs:
```diff
@@ -537,7 +537,7 @@ public void AllSpecialTypeMembers()
 
                 var symbol = comp.GetSpecialTypeMember(special);
-                if (special == SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__DefaultImplementationsOfInterfaces)
+                if (special == SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__DefaultImplementationsOfInterfaces || special == SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__UnmanagedSignatureCallingConvention)
```
>special == SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__UnmanagedSignatureCallingConvention [](start = 131, length = 108)

Consider breaking the line

333fred replied:
I'll do this in a followup if Jared has no other comments.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/WellKnownTypeValidationTests.vb:
```diff
@@ -482,7 +482,7 @@ End Namespace
 
-                If special = SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__DefaultImplementationsOfInterfaces Then
+                If special = SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__DefaultImplementationsOfInterfaces OrElse special = SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__UnmanagedSignatureCallingConvention Then
                     Assert.Null(symbol) ' Not available
```
Consider breaking the line

333fred replied:
I'll do this in a followup if Jared has no other comments.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -184,38 +187,53 @@ static CallingConvention getCallingConvention(Binder typeBinder, FunctionPointer
-                    checkUnmanagedSupport(typeBinder, specifier.GetLocation(), diagnostics, isExtensionError: true);
-                    customModifiers.Add(handleIndividualUnrecognizedSpecifier(specifier, typeBinder, diagnostics));
+                    checkUnmanagedSupport(compilation, specifier.GetLocation(), diagnostics);
+                    var modifier = handleIndividualUnrecognizedSpecifier(specifier, compilation, diagnostics);
```
Nit: you inconsistently use `var` for calling this method.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -184,38 +187,53 @@ static CallingConvention getCallingConvention(Binder typeBinder, FunctionPointer
-                    const string CallingConventionMetadataFormatString = "System.Runtime.CompilerServices.CallConv{0}";
                     string specifierText = specifier.Name.ValueText;
-                    var metadataName = string.Format(CultureInfo.InvariantCulture, CallingConventionMetadataFormatString, specifierText);
+                    if (string.IsNullOrEmpty(specifierText))
```
```suggestion
                    if (specifierText is { Length: > 0})
```

:trollface:

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:
```diff
@@ -599,7 +620,10 @@ private static bool IsCallingConventionModifier(CustomModifier modifier)
             var modifierType = ((CSharpCustomModifier)modifier).ModifierSymbol;
             return (object)modifierType.ContainingAssembly == modifierType.ContainingAssembly.CorLibrary
-                   && modifierType.Name.StartsWith("CallConv", ignoreCase: false, culture: CultureInfo.InvariantCulture)
+                   && modifierType.Name != "CallConv"
+                   && modifierType.Arity == 0
```
```suggestion
                   && modifierType.Arity == 0
                   && modifierType.Name != "CallConv"
```

Do the faster check first.

333fred commented:
I'll address the remaining minor comments in the next PR.
